### PR TITLE
Generate implicit activation events from declared extension contributions

### PR DESCRIPTION
**Pull request by joyceerhl**
For https://github.com/microsoft/vscode/issues/164701

**Authors: @joyceerhl & @alexdima**
This proposed change adds the ability to generate implicit activation events when an extension has certain contributions. This is done by using a registry of implicit activation events generators which are contributed when registering an extension point in our system. We can do another pass to cover all extension points in follow up PRs, this change adds only `onCommand:X` and `onNotebookSerializer:X`. To benefit from caching, we implement this by pushing to the extension's `activationEvents` as soon as the `package.json` file is read (right before doing nls on it). We have explored alternatives, but adding some kind of parallel data structure has proven difficult because the extension descriptions travel quickly to the extension host, even before the extension points are handled.

@sandy081 Please let us know if you see any problems with this approach. I'm not sure anymore if the scanning happens on the renderer process (which loads all the code which deals with activation events i.e. `ImplicitActivationEvents` has participants) or if scanning happens on the shared process (which loads only a subset of our code and over there `ImplicitActivationEvents` might be empty).

### CODE REVIEW

File: src/vs/platform/extensions/common/extensions.ts [hunk 1]
```diff
@@ -202,6 +202,39 @@ export interface IExtensionContributions {
 	readonly notebookRenderer?: INotebookRendererContribution[];
 }
 
+
+export function getImplicitActivationEvents(extensionContributions: IExtensionContributions): string[] {
```
Comment (joyceerhl) [inline src/vs/platform/extensions/common/extensions.ts:206]: Is it worth creating some contribution model to have each main thread contribution consumer expose a mapping from event types to a handler that knows how to take the contribution and convert it to an implicit event? That would ensure that future contribution consumers don't need to find this function in order to opt into implicit events (and of course here I'm making the assumption that these contributions should result in implicit events being generated, which could have unforeseen consequences e.g. extensions getting activated when they don't need to be).

File: src/vs/platform/extensions/common/extensions.ts [hunk 2]
```diff
@@ -202,6 +202,39 @@ export interface IExtensionContributions {
...
+	}
+
+	for (const viewLocation of (Object.values(extensionContributions.views ?? {}))) {
+		for (const view of viewLocation) {
+			activationEvents.push(`onView:${view.id}`);
+		}
+	}
+
+	// TODO: viewContainers, languages, debuggers
+	return activationEvents;
```
Comment (joyceerhl) [inline src/vs/platform/extensions/common/extensions.ts:235]: Not sure whether we should also implicitly activate extensions contributing languages and debuggers, in particular debug has two different kinds of activation events.

**Comment by sandy081** [comment]
@alexdima 

>  I'm not sure anymore if the scanning happens on the renderer process (which loads all the code which deals with activation events i.e. ImplicitActivationEvents has participants) or if scanning happens on the shared process (which loads only a subset of our code and over there ImplicitActivationEvents might be empty).

Scanning happens at both the places. Extensions scanning for extension host (run time) happens on rendered where as extensions scanning for extension management (ui) happens on shared process. This means that package.json content of an extension vary from ui to run time. IMO this shall be ok.

> To benefit from caching, we implement this by pushing to the extension's activationEvents as soon as the package.json file is read (right before doing nls on it). We have explored alternatives, but adding some kind of parallel data structure has proven difficult because the extension descriptions travel quickly to the extension host, even before the extension points are handled.

May I know how the cache gets invalidated if we add a new participant or an existing participant has changed? I am not sure if this needs to be cached as it is dynamic? If not cached, I would prefer to apply these participants here 

https://github.com/microsoft/vscode/blob/a89e89544997a477cf4a9e58015a512a20764f81/src/vs/platform/extensionManagement/common/extensionsScannerService.ts#L262-L280

by introducing a new scan option called `generateImplicitActivationEvents` and the callers pass this option accordingly.

**Comment by joyceerhl** [comment]
This seems not to work in remote windows https://dev.azure.com/vscode/VSCode/_build/results?buildId=78657&view=logs&j=4aebce06-83e9-5552-2127-6f912f36efb7&t=5dac9b77-04f0-5537-3080-6310b0a472f7&l=990 but passes in local windows.

**Comment by joyceerhl** [comment]
From running the test remote resolver locally I found that it looks like we are registering the activation event generators, and at one point we successfully update the manifest with the implicit activation events, but there are then subsequent calls to `updateManifest` which bail out early because it claims there is no registered implicit activation event generator. This is pretty mysterious to me and I'm still investigating.

**Comment by sandy081** [comment]
This is a good catch. In remote scenario, extensions are scanned on the remote server and I am sure activation event generators are not registered on the remote server and hence they are empty and not populated. You have to intercept remote extensions scanning on the client side and update manifests here - 

https://github.com/microsoft/vscode/blob/e3cc677d15dc7a2bb9b2da6332c049361481c85a/src/vs/workbench/services/remote/common/abstractRemoteAgentService.ts#L86-L98

Please note that, you have to update manifest in both methods `scanExtensions` and `scanSingleExtension`. 

IMO this approach of looking at various scan points and intercepting them seems to be not ideal and it's easy to miss if we introduce another scan point. We shall have a single method (service) that extension service calls to get all scanned extensions (local, remote, web) and this method shall scan extensions using respective services and update manifests.

**Comment by joyceerhl** [comment]
Question about caching while implementing a unified scanner--there is the browser [extensionService](https://insiders.vscode.dev/github/microsoft/vscode/blob/0ec04c2b454467586852563bd3a9717bd48763fe/src/vs/workbench/services/extensions/browser/extensionService.ts) and also [electronExtensionService](https://insiders.vscode.dev/github/microsoft/vscode/blob/0ec04c2b454467586852563bd3a9717bd48763fe/src/vs/workbench/services/extensions/electron-sandbox/electronExtensionService.ts), but only the latter uses the `CachedExtensionService`. Is the `CachedExtensionService` what @alexdima was referring to re: caching, and if so, why do we not use it in the browser extension service?

**Review by sandy081 — APPROVED** [review]

**Comment by sandy081** [comment]
@joyceerhl We scan following sources of extensions for extension host

- Desktop Extensions: Scanned for local extension host and caching is triggered [here](https://github.com/microsoft/vscode/blob/94a1ec7999a786336bb0bf6d969401b887126ebe/src/vs/workbench/services/extensions/electron-sandbox/cachedExtensionScanner.ts#L63)
- Remote Extensions: Scanned for remote extension host and caching is triggered [here](https://github.com/microsoft/vscode/blob/94a1ec7999a786336bb0bf6d969401b887126ebe/src/vs/server/node/remoteAgentEnvironmentImpl.ts#L340)
- Web Extensions: Scanned for browser extension host in Web. These are not cached.

**Comment by rgrunber** [comment]
With this change I noticed that when one accesses an extension's [`packageJson`](https://github.com/microsoft/vscode/blob/1.74.0/src/vscode-dts/vscode.d.ts#L7026) property, the `activationEvents` property will also contain those implicit entries (with potential duplicates present).

Admittedly, the package.json contents returned by the API already differed a bit from before (can contain some additional properties) compared to the original file, but based on the field's documentation, it's a bit unclear that this should be expected.

**Comment by sandy081** [comment]
@rgrunber You are right and we are aware of this change. We are planning to fix it in this milestone by not populating these implicit/generated activation events in the package.json.

CC @alexdima

**Pull request merged**